### PR TITLE
Diego authorization

### PR DESCRIPTION
**Pull request by dsoto18**

### CODE REVIEW

File: src/auth/auth.service.ts [hunk 1]
```diff
@@ -34,6 +34,8 @@ export class AuthService {
             throw new HttpException('Email does not exist in database', HttpStatus.NOT_FOUND);
         }
 
+        console.log("validate user function, user: " + user);
```
Comment (pzalep1) [inline src/auth/auth.service.ts:37]: :fire: Console log

File: src/drivers/sendgrid.ts [hunk 2]
```diff
@@ -26,13 +27,45 @@ export class SGService {
...
+      reply_to: {
+        email:"paigezaleppa@gmail.com",
+        name:"Inso API"
+      },
+      templateId: mailinfo.template,
+      dynamicTemplateData: mailinfo.data
+    };
+
+    await this.sgclient.send(email).catch((error) => {
+      console.log(error);
```
Comment (pzalep1) [inline src/drivers/sendgrid.ts:67]: This should throw an Http Error.

File: src/drivers/sendgrid.ts [hunk 3]
```diff
@@ -43,4 +76,39 @@ export class SGService {
...
+  async verifyEmail(user: any){
+    this.sendEmail({
+        name: user.name,
+        username: user.username,
+        contact: user.contact,
+        action: MAIL_DEFAULTS.TEMPLATES.CONFIRM_EMAIL,
+        template: SENDGRID_TEMPLATES.CONFIRM_EMAIL,
+        data: user
+      });
+    console.log(`Email verification sent!`);
```
Comment (pzalep1) [inline src/drivers/sendgrid.ts:88]: :fire: Console log

File: src/drivers/sendgrid.ts [hunk 4]
```diff
@@ -43,4 +76,39 @@ export class SGService {
...
+  async resetPassword(user: any){
+    this.sendEmail({
+      name: user.name,
+      username: user.username,
+      contact: user.contact,
+      action: MAIL_DEFAULTS.TEMPLATES.RESET_PASSWORD,
+      template: SENDGRID_TEMPLATES.PASSWORD_RESET_REQUEST,
+      data: user
+    });
+    console.log(`Password reset request sent!`);
```
Comment (pzalep1) [inline src/drivers/sendgrid.ts:100]: :fire: Console log

File: src/drivers/sendgrid.ts [hunk 5]
```diff
@@ -43,4 +76,39 @@ export class SGService {
...
+  async confirmPassword(user: any){
+    this.sendEmail({
+      name: user.name,
+      username: user.username,
+      contact: user.contact,
+      action: MAIL_DEFAULTS.TEMPLATES.CONFIRM_PASSWORD_RESET,
+      template: SENDGRID_TEMPLATES.PASSWORD_RESET_CONFIRMATION,
+      data: user
+    });
+    console.log(`Password reset confirmation sent!`);
```
Comment (pzalep1) [inline src/drivers/sendgrid.ts:112]: :fire: Console log

File: src/modules/post/post.controller.ts [hunk 6]
```diff
@@ -74,7 +77,7 @@ export class PostController {
       discussionId: new Types.ObjectId(discussionId),
       userId: new Types.ObjectId(req.user.userId),
       date: new Date(),
-      comment_for: new Types.ObjectId(post.comment_for)
+      comment_for: post.comment_for
```
Comment (pzalep1) [inline src/modules/post/post.controller.ts:80]: Does this save as an Objectid or a string?

File: src/modules/user/user.controller.ts [hunk 7]
```diff
@@ -80,7 +89,33 @@ export class UserController {
...
+    if(!foundUser){ 
+      throw new HttpException("No account associated with email: " + email, HttpStatus.NOT_FOUND); 
+    }
+    const userPasswordRequest = {
+      name: foundUser.f_name + ' ' + foundUser.l_name, 
+      username: foundUser.username, 
+      contact: email 
+    }
+    this.sendPasswordResetRequest(userPasswordRequest);
+    console.log('Password reset request has been sent to email: ' + email);
```
Comment (pzalep1) [inline src/modules/user/user.controller.ts:117]: :fire: Console log

File: src/modules/user/user.controller.ts [hunk 8]
```diff
@@ -176,29 +211,66 @@ export class UserController {
     return 'User Updated';
   }
 
+  @Patch('password-reset')
+  async resetPassword(
+    @Body() newPassword:{ password: string, confirmPassword: string },
+    @Query('ota') ota: string
+  ){
+    console.log(ota);
```
Comment (pzalep1) [inline src/modules/user/user.controller.ts:219]: :fire: Console log

File: src/modules/user/user.controller.ts [hunk 9]
```diff
@@ -176,29 +211,66 @@ export class UserController {
...
+    }
+    validatePassword(newPassword.password);
+    this.verifyPasswordResetToken(ota, newPassword.password);
+  }
 
-//////////////////////////////////
+  //**  Uses SendGrid to send email, function is performed at the end of user registration (POST USER ROUTE) */
+  async sendEmailVerification(user: any){
+    const ota = await generateCode(user.contact);
+    console.log(ota);
```
Comment (pzalep1) [inline src/modules/user/user.controller.ts:230]: :fire: Console log

File: src/modules/user/user.controller.ts [hunk 10]
```diff
@@ -176,29 +211,66 @@ export class UserController {
...
+    return  await this.sgService.verifyEmail({...user, link: 'http://localhost:3000/email-verified?ota=' + ota.code});
+  }
 
-/** removes contacts with delete boolean marker as true, for patch route */
-  public removeUnwantedContacts(array: ContactEditDTO[]){
+  async verifyEmailToken(ota: string){
+    const code = await decodeOta(ota);
+    //const queryData = {"email": code.data, "verified": true, "primary": true}
```
Comment (pzalep1) [inline src/modules/user/user.controller.ts:236]: If we're not using the commented code let's remove it

File: src/modules/user/user.controller.ts [hunk 11]
```diff
@@ -176,29 +211,66 @@ export class UserController {
...
+    //console.log(await this.userModel.aggregate( [ {"$project": { "matchedIndex": { "$indexOfArray": [ "$contact", queryData]}}}]));
+    
+    const checkVerified = await this.userModel.findOne({'contact.email' : code.data});
+    console.log(checkVerified);
```
Comment (pzalep1) [inline src/modules/user/user.controller.ts:240]: :fire: Console log

File: src/modules/user/user.controller.ts [hunk 12]
```diff
@@ -176,29 +211,66 @@ export class UserController {
...
+    const arr = checkVerified.contact;
+    for(let e of arr){
+      if (e.email == code.data && e.verified == true){
+        console.log('Email has already been verified');
```
Comment (pzalep1) [inline src/modules/user/user.controller.ts:244]: :fire: Console log

File: src/modules/user/user.controller.ts [hunk 13]
```diff
@@ -176,29 +211,66 @@ export class UserController {
...
+    }
 
-    var contactsToDelete = array.filter(function (e) {
-      return e.delete == true;
-    }); // new array of elements to remove from current contacts
-    var contactsToKeep = array.filter(function (e) {
-      return e.delete != true;
-    }); // new array of new elements to add to current contacts
+    //if( checkVerified.contact)
+    //console.log(checkVerified);
```
Comment (pzalep1) [inline src/modules/user/user.controller.ts:250]: If this isn't used lets' remove it

File: src/modules/user/user.controller.ts [hunk 14]
```diff
@@ -176,29 +211,66 @@ export class UserController {
...
 
-    for (var i = 0; i < contactsToDelete.length; i++) {
-    this.userModel.updateMany(
-      {},
-      { $pull: { contact : {email: contactsToDelete[i].email} } }
-      )
-    }
+    await this.userModel.findOneAndUpdate({'contact.email': code.data}, { $set: {'contact.$.verified': true}});
+    console.log('Email verified!');
```
Comment (pzalep1) [inline src/modules/user/user.controller.ts:253]: :fire: Console log

File: src/modules/user/user.controller.ts [hunk 15]
```diff
@@ -176,29 +211,66 @@ export class UserController {
...
+  }
 
-    return contactsToKeep;
+  async sendPasswordResetRequest(user: any){
+    const ota = await generateCode(user.contact);
+    console.log(ota);
```
Comment (pzalep1) [inline src/modules/user/user.controller.ts:258]: :fire: Console log

**Review by pzalep1 — CHANGES_REQUESTED** [review]

File: src/auth/auth.service.ts (hunk 1 above)
Comment (dsoto18) [reply src/auth/auth.service.ts:37]: Done, I had a lot of these to test if the information was getting passed through correctly.

File: src/drivers/sendgrid.ts (hunk 3 above)
Comment (dsoto18) [reply src/drivers/sendgrid.ts:88]: Done.

File: src/drivers/sendgrid.ts (hunk 4 above)
Comment (dsoto18) [reply src/drivers/sendgrid.ts:100]: Done.

File: src/drivers/sendgrid.ts (hunk 5 above)
Comment (dsoto18) [reply src/drivers/sendgrid.ts:112]: Done.

File: src/modules/post/post.controller.ts (hunk 6 above)
Comment (dsoto18) [reply src/modules/post/post.controller.ts:80]: My thinking was that 'new Types.ObjectId()' would always save as a MongoId, I can check it though.

File: src/modules/user/user.controller.ts (hunk 8 above)
Comment (dsoto18) [reply src/modules/user/user.controller.ts:219]: Done.

File: src/modules/user/user.controller.ts (hunk 7 above)
Comment (dsoto18) [reply src/modules/user/user.controller.ts:117]: Done.

File: src/modules/user/user.controller.ts (hunk 9 above)
Comment (dsoto18) [reply src/modules/user/user.controller.ts:230]: Done.

File: src/modules/user/user.controller.ts (hunk 10 above)
Comment (dsoto18) [reply src/modules/user/user.controller.ts:236]: Yes, just removed it all.

File: src/modules/user/user.controller.ts (hunk 11 above)
Comment (dsoto18) [reply src/modules/user/user.controller.ts:240]: Done.

File: src/modules/user/user.controller.ts (hunk 12 above)
Comment (dsoto18) [reply src/modules/user/user.controller.ts:244]: Done.

File: src/modules/user/user.controller.ts (hunk 13 above)
Comment (dsoto18) [reply src/modules/user/user.controller.ts:250]: It is removed.

File: src/modules/user/user.controller.ts (hunk 14 above)
Comment (dsoto18) [reply src/modules/user/user.controller.ts:253]: Done.

File: src/modules/user/user.controller.ts (hunk 15 above)
Comment (dsoto18) [reply src/modules/user/user.controller.ts:258]: Done.

File: src/modules/post/post.controller.ts (hunk 6 above)
Comment (dsoto18) [reply src/modules/post/post.controller.ts:80]: It is either an ObjectId or it is null and it saves with the field empty.

**Pull request merged**